File: 2024/day11.hs
import qualified Data.Map.Strict as Map
import Data.List.Split (wordsBy)
import Text.Read (readMaybe)
import System.IO (hFlush, stdout)

type Cache = Map.Map (Int, Integer) Integer

blink :: Int -> Integer -> Cache -> (Integer, Cache)
blink 0 _ cache = (1, cache)
blink times stone cache =
    case Map.lookup state cache of
        Just cached -> (cached, cache)
        Nothing ->
            let (result, newCache) = if stone == 0
                    then blink (times - 1) 1 cache
                    else
                        let digitCount = floor (logBase 10 (fromIntegral stone)) + 1 :: Int
                        in if digitCount `mod` 2 == 0
                            then
                                let divisor = 10 ^ (digitCount `div` 2)
                                    (part1, cache1) = blink (times - 1) (stone `div` divisor) cache
                                    (part2, cache2) = blink (times - 1) (stone `mod` divisor) cache1
                                in (part1 + part2, cache2)
                            else blink (times - 1) (stone * 2024) cache
                updatedCache = Map.insert state result newCache
            in (result, updatedCache)
  where
    state = (times, stone)

main :: IO ()
main = do
    input <- getLine
    let stones = map readInteger (words input)
    let timesList = [25, 75]
    let initialCache = Map.empty

    mapM_ (\times ->
        let (count, _) = foldl (\(acc, cache) stone ->
                let (res, newCache) = blink times stone cache
                in (acc + res, newCache)
              ) (0, initialCache) stones
        in putStrLn $ show count
      ) timesList
  where
    readInteger :: String -> Integer
    readInteger s = case readMaybe s of
        Just n  -> n
        Nothing -> error "Failed to read input"

File: 2024/day11.rs
use std::collections::HashMap;
use std::io::{self, BufRead};

fn blink(times: u8, stone: i64, cache: &mut HashMap<u64, i64>) -> i64 {
    if times == 0 {
        return 1;
    }

    let key = ((times as u64) << 56) | (stone as u64);
    if let Some(&result) = cache.get(&key) {
        return result;
    }

    let result = if stone == 0 {
        blink(times - 1, 1, cache)
    } else {
        let digit_count = (stone as f64).log10().floor() as u32 + 1;
        if digit_count % 2 == 0 {
            let divisor = 10_i64.pow(digit_count / 2);
            let lhs = stone / divisor;
            let rhs = stone % divisor;
            blink(times - 1, lhs, cache) + blink(times - 1, rhs, cache)
        } else {
            blink(times - 1, stone * 2024, cache)
        }
    };

    cache.insert(key, result);
    result
}

fn main() {
    let stdin = io::stdin();
    let mut input = String::new();

    if let Err(err) = stdin.lock().read_line(&mut input) {
        eprintln!("read_line: {}", err);
        return;
    }

    let stones: Vec<i64> = input
        .trim()
        .split_whitespace()
        .map(|s| s.parse().unwrap())
        .collect();

    for &times in &[25, 75] {
        let mut cache = HashMap::new();
        let count: i64 = stones.iter().map(|&stone| blink(times, stone, &mut cache)).sum();
        println!("{}", count);
    }
}
